src/devices/Mcp23xxx/Mcp23xxx.cs

```diff
@@ -314,20 +304,16 @@ public void SetPinMode(int pinNumber, PinMode mode)
         public PinValue Read(int pinNumber)
         {
             ValidatePin(pinNumber);
-            Span<PinValuePair> values = stackalloc PinValuePair[1];
-            values[0] = new PinValuePair(pinNumber, default);
+            Span<PinValuePair> values = stackalloc PinValuePair[] { new PinValuePair(pinNumber, default) };
             Read(values);
             return values[0].PinValue;
         }
 
         public void Read(Span<PinValuePair> pinValues)
         {
-            ushort pins = 0;
-            foreach (PinValuePair pair in pinValues)
-            {
-                ValidatePin(pair.PinNumber);
-                pins |= (ushort)(1 << pair.PinNumber);
-            }
+            (uint pins, _) = new PinVector32(pinValues);
+            if ((pins >> PinCount) > 0)
+                ThrowBadPin(nameof(pinValues));
```

Comment:
I'd recommend BadPin return the exception and do `throw BadPin(...)` here - the stack looks better (no strong opinion)

Reply:
While I didn't really need it in this particular case, in some cases it is important to have the throw come from another method. Throws prevent the JIT from inlining methods.